### ISSUE TITLE
Deprecate string datetimes support

### ISSUE THREAD

**Issue by theofidry** [nelmio/alice#495]
Looking at the TypehintChecker, it allows us to do two things:

- transforming string dates to dates
- looking for an existing instance (via the persister)

There is three reasons why I think we could remove it:

1. The first point, which is the most useful and common IMO, could be done with a provider instead
2. The second point should not be needed as if you use `@dummy`, you will get an instance right away so the typechecker is not required there. The only valid use case is if you are using a string which refers to an ID handled by a persister, use case irrelevant in v3.
3. It makes use of the reflection for every property, it is slow as hell

The reason why I have mixed feelings about this one is that removing it completely in v3 would introduce another non negligible user-land BC break.

An alternative, less extreme, would be to have a "workaround" in v3 to ensure BC. This workaround would be a less performant than the regular system, would occur only for the first use case (dates). Besides this workaround, we could deprecate the date feature in both 2.x and 3.x and provide a custom provider as a replacement. This way there is not user-land BC break and the performance issue will impact only users which are using the deprecate feature.

WDYT @tshelburne?

**Comment by tshelburne** [comment]
I can't remember and don't have a moment to dig in right now, but I thought the type hint checker was about instantiating an instance properly via constructor arguments before setting all the props. Might be completely mistaken.

As far as the BC problem, what exactly does it necessarily break? Without digging in, this seems like an "internals" question, not a library interface question.

**Comment by theofidry** [comment]
>I thought the type hint checker was about instantiating an instance properly via constructor arguments before setting all the props.

No you are right except that it also applies to setters. In both cases, it involves check that the value set matches the typehint. The most common use case is for dates:

```php
public function testLoadCoercesDatesForDateTimeHints()
    {
        $res = $this->loadData([
            self::GROUP => [
                'group0' => [
                    'creationDate' => '2012-01-05',
                ],
                'group1' => [
                    'creationDate' => '<unixTime()>',
                ],
            ],
        ]);

        $this->assertInstanceOf('DateTime', $res['group0']->getCreationDate());
        $this->assertEquals('2012-01-05', $res['group0']->getCreationDate()->format('Y-m-d'));
        $this->assertInstanceOf('DateTime', $res['group1']->getCreationDate());
    }
```

>As far as the BC problem, what exactly does it necessarily break? Without digging in, this seems like an "internals" question, not a library interface question.

Right now you can do the following:

```yaml
Group:
  group0:
    creationDate: '2012-01-05'
  group1:
    creationDate: '<unixTime()>'
```

Without any change, this won't work in 3.x because the setter `setCreationDate()` expects a `\DateTime` object. This works in 2.x thanks to the typehint checker right now, hence my suggestions from above.

**Comment by tshelburne** [comment]
Ah, I see. Personally, I don't think it should be Alice's responsibility to turn what reads as a string into a date time object - that's what Faker is for. I'm not opposed to leaving it in there, but it does seem like a high price to pay for a feature that is only arguably useful. 

I think we should introduce an official migration path to use `creationDate: <date('2012-01-05')>`, or something like that. I'd honestly be surprised if that doesn't already exist.

**Comment by theofidry** [comment]
True, there is already https://github.com/fzaninotto/Faker#fakerproviderdatetime. So we agree to deprecate it in 2.x and remove it for 3.x?

**Comment by tshelburne** [comment]
Yes

**Comment by Soullivaneuh** [comment]
> True, there is already https://github.com/fzaninotto/Faker#fakerproviderdatetime.

This produce a **random** `DateTime` instance. For `2012-01-05` on your example, it's a fixed date.

This is not the same, is this?

**Comment by theofidry** [comment]
true, but you can always do `<(new \DateTimeImmutable('2012-01-05'))>`

**Comment by theofidry** [comment]
Closed via #559
